_posts/2020-11-12-symbolic-execution-with-ds-test.md

```diff
@@ -0,0 +1,429 @@
+
+To illustrate the differences between the test types, consider the following example:
+
+```solidity
+contract Test is DSTest {
+    function test_associativity() public {
+        assertEq(
+            uint((1 + 2) + 3),
+            uint(1 + (2 + 3))
+        );
+    }
+
+    function test_associativity_fuzz(uint x, uint y, uint z) public {
+        assertEq(
+            (x + y) + z,
+            x + (y + z)
+        );
+    }
+
+    function prove_associativity(uint x, uint y, uint z) public {
+        assertEq(
+          (x + y) + z,
+          x + (y + z)
+        );
+    }
+}
+```
+
+- `test_associativity` will run a single test case with the concrete values of `1`, `2`, and `3`.
+- `test_associativity_fuzz` will be executed many times (100 by default), with randomly generated values for `x`, `y`, and `z`.
+- `prove_associativity` will be symbolically executed, with `x`, `y`, and `z` represented as symbolic variables.
+
+Finally, it is possible to manipulate the execution environment (e.g. timestamp) from within
+`ds-test` using by using hevm "[cheat codes](https://github.com/dapphub/dapptools/tree/master/src/hevm#cheat-codes)"
+
+## Finding Counterexamples
+
+Let's look at a more complex example! Consider the following token contract:
+
+```solidity
+pragma solidity ^0.6.12;
+
+contract SafeMath {
+    function add(uint x, uint y) internal pure returns (uint z) {
+        require((z = x + y) >= x, "overflow");
+    }
+    function sub(uint x, uint y) internal pure returns (uint z) {
+        require((z = x - y) <= x, "underflow");
+    }
+}
+
+contract Token is SafeMath {
+    uint256 public totalSupply;
+    mapping (address => uint) public balanceOf;
+
+    constructor(uint _totalSupply) public {
+        totalSupply           = _totalSupply;
+        balanceOf[msg.sender] = _totalSupply;
+    }
+
+    function transfer(address dst, uint amt) public {
+        balanceOf[msg.sender] = sub(balanceOf[msg.sender], amt);
+        balanceOf[dst]        = add(balanceOf[dst], amt);
+    }
+}
+```
+
+We can write a test that asserts our expected behaviour for the transfer function as follows:
+
+```solidity
+contract TestToken is DSTest, SafeMath {
+    function prove_transfer(address dst, uint amt) public {
+        Token token = new Token(uint(-1));
+
+        uint preBalThis = token.balanceOf(address(this));
+        uint preBalDst  = token.balanceOf(dst);
+
+        token.transfer(dst, amt);
+
+        // balance of this has been reduced by `amt`
+        assertEq(sub(preBalThis, token.balanceOf(address(this))), amt);
+
+        // balance of dst has been increased by `amt`
+        assertEq(sub(token.balanceOf(dst), preBalDst), amt);
+    }
+}
+```
+
+Interestingly, if we run this test with `dapp test -v`, it fails with the following output:
+
+```
+Running 1 tests for src/Solidity.t.sol:TestERC20
+[FAIL] prove_transfer(address,uint256)
+
+Failure: prove_transfer(address,uint256)
+
+  Counter Example:
```

Comment:
```suggestion
  Counterexample:
```
More common?

Reply:
hmmm yeah this has been bugging me a little as well actually, will change in hevm.